### ISSUE TITLE
Streaming interface

### ISSUE THREAD

**Issue by maxtaco** [keybase/triplesec#15]
For command-line file encyptor.

**Comment by dominictarr** [comment]
or you could use multiple headers, i.e. a framed protocol. 
for example like @calvinmetcalf's https://github.com/calvinmetcalf/hmac-stream/ or my
https://github.com/dominictarr/pull-box-stream

**Comment by SparkDustJoe** [comment]
miniLock suffers from this same problem of there being data in the header at the top of the file that isn't known until the cipher-text has been created "somewhere" in full.  Changing the header to be "magic bytes" + version + salt + IV + CT + signatures at the end, in either a raw byte stream or JSON, would do it for TripleSec on the encryption side as far as streaming the output.  For decryption having the keys up front and doing some calculations as the file came down and then doing the decryption when enough has been downloaded (since this protocol is layered, and the IV's are posted up-front), and then calculating the signatures as a last process would be faster too.

**Comment by SparkDustJoe** [comment]
think of it like putting down three layers of tape.  The bottom-most layer doesn't have to be fully placed for the next layer to start being applied on top, and then the next layer of top of that as the lower two are placed ahead.  All of the IV's are known ahead of time, so as one layer is encrypted, past a certain point the next layer with the IV affixed to the beginning can be encrypted, and finally the last.  At the end the signatures are generated on the last layer of "tape" (cipher-text), and sent as the last output.

**Comment by SparkDustJoe** [comment]
AES and Twofish have a 128 bit block width, but XSalsa20 has a 512 block width (which is exactly 4 times the other two but you have to factor in the 128 bit IV for TwoFish and the 160 bit IV for Salsa). Decrypting the first 7 blocks of data from the beginning (which strips off the AES layer) gives you enough to start the process.

**Comment by calvinmetcalf** [comment]
AES in CTR mode, xsalsa and twofish are all streaming so the blocksize is irreverent, the bigger issue is that you don't want to decrypt unverified data so buffering is necessary to get all the cipher text so it can be verified before decryption so having streaming decrypting is pretty much unnecessary since you already have things in memory.  

That being said doing a framing approach where you chunk your message and encrypt each of them separately would be useful except there is a tremendous amount of per message scrypt overhead since you do `(password + salt) -> key` 5 separate times having a method to reuse the keys and create new messages quickly without having to re derive the keys might be more practical for streaming.  Reusing the salt might be one way to do it

**Comment by SparkDustJoe** [comment]
It's not that '<(password + salt) -> key>' is run 5 times, the same run produces a long byte array that CONTAINS 5 keys (3 for encryption, 2 for signatures, all done at once). But I see your point with trying to decrypt something that might be invalid to begin with, no sense in doing all that work if the stream is corrupt.  But if the signatures are at the end, then you still could at least feed the cipher-text stream into the hashes as it comes down, verify, and THEN decrypt from memory.  Not sure how much time this saves with smaller files, but larger ones it could make a difference.   

In miniLock (which only uses one encryption pass), there is a chunk number worked into the nonce of the files as they are encrypted with XSalsa20-Poly1305 (which the Poly1305 part isn't used here).  Something to that effect could be used where the MASTER key is generated from '<password + salt>'  but then is HMAC'd with '<chunk number + IV>' or '<chunk number XOR IV>', so that the keys for each chunk are dependent on which chunk you are talking about and still tied to the scheme so that forgery is still incredibly difficult.  Each chunk doesn't need authentication per se if the WHOLE of the cipher-text is authenticated on download.

As far as the CTR mode, there is a certain amount of information you have to have first to get the process started, so maybe it's less than the 7 blocks that I stated above, but there is still some work to be done peeling away layers to get to the IV's of the "lower" functions on decrypt if you wanted to do them concurrently.

**Comment by calvinmetcalf** [comment]
> But I see your point with trying to decrypt something that might be invalid to begin with, no sense in doing all that work if the stream is corrupt.

more then that there are attacks that work by passing in invalid data and depending on the error being able to gain information

**Comment by SparkDustJoe** [comment]
Since the CTR mode as no authentication inherent to it (unlike with Poly1305), the system probably wouldn't know until the very end (or if the signatures failed) unless it was doing some kind of "header" or "leader" check on the decrypted data, so it would depend on how TripleSec was used in a larger system, but yeah that's also a very valid point.

**Comment by calvinmetcalf** [comment]
Well of triplesec does the authentication then there is no reason to add
multiple levels of it latter in the scheme

**Comment by dodekeract** [comment]
Any news on this? It would be amazing to have a secure, streamable cipher.

**Comment by maxtaco** [comment]
We are working on one now with saltpack! See our work over in
node-saltpack.

**Comment by dodekeract** [comment]
@maxtaco If I understand this correctly, it's using asymmetric cryptography. That's nice for messaging, but unfortunately isn't good for en- and decrypting files, without sharing them.

**Comment by maxtaco** [comment]
You're right of course! It's not currently on our road map.  A hack is to take a 32-byte secret key and to treat that as a Curve25519 private key, and then encrypt to yourself.  But it's inelegant and adds an additional assumption/dependence, that Curve25519 isn't broken....
